wcfsetup/install/files/lib/system/event/EventHandler.class.php

```diff
@@ -185,48 +156,90 @@ public function fireAction($eventObj, $eventName, array &$parameters = []) {
-								throw new ImplementationException($eventListener->listenerClassName, IParameterizedEventListener::class);
-							}
-						}
-						
-						$object = new $eventListener->listenerClassName;
-						$this->listenerObjects[$eventListener->listenerClassName] = $object;
-					}
-					
-					$this->actionsObjects[$name][$eventListener->listenerClassName] = $object;
-				}
+				if (isset($this->actionsObjects[$name][$eventListener->listenerClassName])) continue;
+				
+				$object = $this->createEventObject($eventListener);
+				if ($object !== null) $this->actionsObjects[$name][$eventListener->listenerClassName] = $object;
 			}
 		}
 		
 		// execute actions
 		foreach ($this->actionsObjects[$name] as $actionObj) {
-			if ($actionObj instanceof IParameterizedEventListener) {
-				$actionObj->execute($eventObj, $className, $eventName, $parameters);
+			$this->callEvent($eventObj, $actionObj, $className, $eventName, $parameters);
+		}
+	}
+	
+	/**
+	 * Create the event object and check if this implements the interfaces
+	 * 
+	 * @param EventListener $eventListener
+	 *
+	 * @return IEventListener|IParameterizedEventListener|IDynamicInvocationEventListener|null
+	 * @throws SystemException
+	 */
+	protected function createEventObject($eventListener) {
+		if ($eventListener->validateOptions() && $eventListener->validatePermissions()) {
+			// get class object
+			if (isset($this->listenerObjects[$eventListener->listenerClassName])) {
+				$object = $this->listenerObjects[$eventListener->listenerClassName];
+			} else {
+				$object = null;
+				// instance action object
+				if (!class_exists($eventListener->listenerClassName)) {
+					throw new SystemException("Unable to find class '" . $eventListener->listenerClassName . "'");
+				}
+				if (!is_subclass_of($eventListener->listenerClassName, IParameterizedEventListener::class)) {
+					// legacy event listeners
+					if (!is_subclass_of($eventListener->listenerClassName, IEventListener::class)) {
+						throw new ImplementationException($eventListener->listenerClassName, IParameterizedEventListener::class);
+					}
+				}
 				
-				if (!is_array($parameters)) {
-					throw new SystemException("'".get_class($actionObj)."' breaks the '\$parameters' array!");
+				$object = new $eventListener->listenerClassName;
+				$this->listenerObjects[$eventListener->listenerClassName] = $object;
+			}
+			return $object;
+		}
+		
+		return null;
+	}
+	
+	/**
+	 * This function calls the event listener
+	 * Function to reduce duplicate code in the fire action between inherited and not inherited actions
+	 *
+	 * @param object $eventObj
+	 * @param object $actionObj
+	 * @param string $className
+	 * @param string $eventName
+	 * @param array &$parameters
+	 *
+	 * @throws SystemException
+	 * @throws \BadMethodCallException
+	 */
+	protected function callEvent($eventObj, $actionObj, $className, $eventName, &$parameters) {
+		if ($actionObj instanceof IDynamicInvocationEventListener) {
```

Comment:
Why would we need this special handling here? Wouldn't an abstract parent class that performs the dispatch work?